packages/lesswrong/components/posts/PostsPage/PostsPageCrosspostWrapper.tsx

```diff
@@ -0,0 +1,68 @@
+import React, { createContext, useContext } from "react";
+import { Components, registerComponent } from "../../../lib/vulcan-lib";
+import { useSingle, UseSingleProps } from "../../../lib/crud/withSingle";
+import { isMissingDocumentError, isOperationNotAllowedError } from "../../../lib/utils/errorUtil";
+import { useCrosspostApolloClient } from "../../hooks/useCrosspostApolloClient";
+
+type PostType = PostsWithNavigation | PostsWithNavigationAndRevision;
+
+export type CrosspostContext = {
+  hostedHere: boolean,
+  localPost: PostType,
+  foreignPost?: PostType,
+  combinedPost?: PostType,
+}
+
+const crosspostContext = createContext<CrosspostContext | null>(null);
+
+export const useCrosspostContext = () => useContext(crosspostContext);
+
+const PostsPageCrosspostWrapper = ({post, refetch, fetchProps}: {
+  post: PostType,
+  refetch: () => Promise<void>,
+  fetchProps: UseSingleProps<"PostsWithNavigation"|"PostsWithNavigationAndRevision">,
+}) => {
+  const apolloClient = useCrosspostApolloClient();
+  const { document, loading, error } = useSingle<"PostsWithNavigation"|"PostsWithNavigationAndRevision">({
+    ...fetchProps,
+    documentId: post.fmCrosspost!.foreignPostId!,
```

Comment:
I spent a bit of time looking into it and it would be non-trivial to modify our type code-gen to allow us to properly represent the actual invariants for `fmCrosspost`, which, as far as I can tell, are as follows:
```
type NotCrosspost = {
  isCrosspost: false;
};

type HostedHereCrosspost = {
  isCrosspost: true;
  hostedHere: true;
}

type HostedElsewhereCrosspost = {
  isCrosspost: true;
  hostedHere: false;
  foreignPostId: string;
}

type FMCrosspost = NotCrosspost | HostedElsewhereCrosspost | HostedElsewhereCrosspost;
```

But in `PostsPageWrapper`, we're only checking for `isCrosspost: true`, which leaves open the possibility that `foreignPostId` won't have a value, if it's hosted "here".  (I could be confused/mistaken about the logic with the post wrappers.)

We could eliminate the usage of `!` by narrowing the type of the `post` param we accept here, and checking it (e.g. via type guard) in `PostsPageWrapper` instead of just relying on `if (post.fmCrosspost?.isCrosspost)`.

Reply:
The only time that `isCrosspost` is `true` and `foreignPostId` is `false` is if the post is still a draft and hasn't been crossposted yet. If the crosspost has been created then `foreignPostId` will always point to the post on the other site, whether or not it's also `hostedHere`.

I've just updated PostsPageWrapper to add a PostWithForeignId type and a type guard - it should be a bit nicer now!